### DISCUSSION TITLE
Utilizing multiple gates for the same feature flag

### DISCUSSION THREAD

**Discussion by slants3** [jnunemaker/flipper#738] [Q&A] [answered]
Hi! I'm currently working on a project where I'd like to use multiple gates for one feature flag.

Basically I'd like to pass in a condition for customers here in a `Group` block and then have the `PercentageOfActors` divide the qualifying customers by 50% so only half see the new feature. What I tried was to create a group block and then also turned on 50% of actors in the UI in a way to kind of gimmick around it:
```
Flipper.register(:qualified_customers) do |actor|
  actor.condition_1? && actor.condition_2? && Flipper::Gates::PercentageOfActors.open?(actor, :flag_for_qualified_customers)
end
```

But once a customer satisfies condition1 and condition2, they are automatically enabled via the flag and the last bit doesn't work. It seems like `open?` isn't a valid method and I'm copying this to demonstrate what I'm trying to do.

I see in a comment [here](https://github.com/jnunemaker/flipper/issues/71#issuecomment-106310962) that Flipper basically enables the flag once any of the gates are enabled so I'm unsure what I'm looking for here is possible.

**Comment by bkeepers** [comment, marked as answer]
Hey @slants3, just to rephrase to make sure I understand: you want to enable a feature for half of the users in the `:qualified_customers` group?

One way to approach this would be to have the group then call `Flipper.enabled?` for a dynamically generated feature name:

```ruby
Flipper.register(:qualified_customers) do |actor, context|
  actor.condition_1? &&
  actor.condition_2? &&
  Flipper.enabled?("#{context.feature_name}_for_qualified_customers", actor)
end
```

Now when you enable a feature for the `:qualified_customers` group, you also need to enable the secondary feature:

```ruby
Flipper.enable_group(:search, :qualified_customers)

# Enable for 50% of qualified customers
Flipper.enable_percentage_of_actors(:search_for_qualified_customers, 50)

# Enable for qualified customers that are also VIPs
Flipper.enable_group(:search_for_qualified_customers, :vip)

# Enable for all qualified customers
Flipper.enable_group(:search_for_qualified_customers)

# Enable for everyone
Flipper.enable(:search)
```

This approach gives you full runtime control over which users get the feature so you don't have to redeploy to change the group requirements.

Does that achieve what you wanted?

**Reply by slants3** [reply]
Hi @bkeepers, I believe it does!

I am a bit confused - I was under the impression `actor` represents any model and does some calculation based on the model and its ID to calculate if it is enabled via percentage of actors. How would Flipper have an understanding of the total number of customers that qualify for `condition_1` and `condition_2` to divide them appropriately for percentage of actors?
